sw/device/tests/i2c_host_clock_stretching_test.c

```diff
@@ -0,0 +1,255 @@
+
+  // Reset the delay to disable clock stretching.
+  {
+    uint8_t write_buffer[2] = {kRxDelayReg, 0};
+    TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(write_buffer),
+                            write_buffer, true));
+  }
+
+  TRY(dif_i2c_reset_rx_fifo(&i2c));
+
+  return OK_STATUS();
+}
+
+static status_t tx_stretch(void) {
+  enum { kTimeoutMillis = 1, kTxSize = 63 };
+
+  // Set the delay that will make the target to stretch the clock.
+  {
+    uint8_t write_buffer[2] = {kTxDelayReg, kTimeoutMillis};
+    TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(write_buffer),
+                            write_buffer, true));
+  }
+
+  // Init buffer with random data.
+  uint8_t rnd_data[kTxSize];
+  for (int i = 0; i < sizeof(rnd_data); ++i) {
+    uint32_t rand;
+    TRY(rv_core_ibex_testutils_get_rnd_data(&rv_core_ibex, 1000, &rand));
+    rnd_data[i] = rand & 0xFF;
+  }
+  uint8_t write_buffer[kTxSize + 1];
+  write_buffer[0] = kCache63BitsReg;
+  memcpy(&write_buffer[1], rnd_data, sizeof(rnd_data));
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(write_buffer), write_buffer,
+                          false));
+  // The transmission may take a long time due to the clock stretching.
+  busy_spin_micros((kTimeoutMillis * kTxSize) * 1000);
+
+  uint8_t reg = kCache63BitsReg;
+  uint8_t read_data[kTxSize] = {0};
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(reg), &reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, sizeof(read_data), read_data));
+  TRY_CHECK_ARRAYS_EQ(read_data, rnd_data, sizeof(read_data));
+
+  // Reset the delay to disable clock stretching.
+  {
+    uint8_t write_buffer[2] = {kTxDelayReg, 0};
+    TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(write_buffer),
+                            write_buffer, true));
+  }
+
+  return OK_STATUS();
+}
+
+static status_t test_init(void) {
+  mmio_region_t base_addr =
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR);
+
+  TRY(dif_rv_core_ibex_init(base_addr, &rv_core_ibex));
+
+  base_addr = mmio_region_from_addr(TOP_EARLGREY_I2C0_BASE_ADDR);
+  TRY(dif_i2c_init(base_addr, &i2c));
+
+  base_addr = mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR);
+  TRY(dif_pinmux_init(base_addr, &pinmux));
+  TRY(i2c_testutils_connect_i2c_to_pinmux_pins(&pinmux, 0));
+  TRY(dif_i2c_host_set_enabled(&i2c, kDifToggleEnabled));
+
+  base_addr = mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR);
+  TRY(dif_rv_plic_init(base_addr, &plic));
+
+  rv_plic_testutils_irq_range_enable(&plic, kHart,
+                                     kTopEarlgreyPlicIrqIdI2c0FmtThreshold,
+                                     kTopEarlgreyPlicIrqIdI2c0HostTimeout);
+
+  // Enable the external IRQ at Ibex.
+  irq_global_ctrl(true);
+  irq_external_ctrl(true);
+
+  return OK_STATUS();
+}
+
+bool test_main(void) {
+  CHECK_STATUS_OK(test_init());
+
+  dif_i2c_speed_t speeds[] = {kDifI2cSpeedStandard/*, kDifI2cSpeedFast,
```

Comment:
Do you intend eventually to run this test all all three speeds?

Reply:
Yes, I commented out during the debug and forgot to revert. Fixed.